packages/core/src/types.tsx

```diff
@@ -648,6 +648,18 @@ export type NavigationContainerEventMap = {
   };
 };
 
+type ParamListRoute<ParamList extends ParamListBase> = {
+  [RouteName in keyof ParamList]: ParamList[RouteName] extends NavigatorScreenParams<
+    infer T extends ParamListBase
+  >
+    ? ParamListRoute<T>
+    : Route<Extract<RouteName, string>, ParamList[RouteName]>;
+}[keyof ParamList];
+
+type MaybeParamListRoute<ParamList extends {}> = ParamList extends ParamListBase
```

Comment:
Less than sure on the naming for this one. Need a second type to manage the case were `ParamList` does not extend `ParamListBase`, as the restriction on the ref is `extends {}`.

Would need the name to signify that it might be a "well typed" `Route` or it could be `Route<string>` depending on what you pass.